src/storage/src/object/disk.rs

```diff
@@ -0,0 +1,495 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::DefaultHasher;
+use std::fs::File;
+use std::hash::{Hash, Hasher};
+use std::io::{Error, ErrorKind};
+use std::os::unix::fs::FileExt;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use bytes::Bytes;
+use futures::future::join_all;
+use tokio::io::AsyncWriteExt;
+use tokio::sync::mpsc::UnboundedSender;
+
+use crate::hummock::{CachableEntry, HummockError, LruCache};
+use crate::object::{BlockLocation, ObjectError, ObjectMetadata, ObjectResult, ObjectStore};
+
+pub(super) mod utils {
+    use std::fs::Metadata;
+    use std::path::Path;
+
+    use tokio::fs::{create_dir_all, OpenOptions};
+    use tokio::task::spawn_blocking;
+
+    use super::OpenReadFileHolder;
+    use crate::object::{ObjectError, ObjectResult};
+
+    pub async fn ensure_file_dir_exists(path: &Path) -> ObjectResult<()> {
+        if let Some(dir) = path.parent() {
+            // `create_dir_all` will not error even if the directory already exists.
+            create_dir_all(dir).await.map_err(|e| {
+                ObjectError::disk(format!("unable to create dir: {:?}.", dir.to_str(),), e)
+            })?;
+        }
+        Ok(())
+    }
+
+    pub async fn open_file(
+        path: &Path,
+        enable_read: bool,
+        enable_write: bool,
+        create_new: bool,
+    ) -> ObjectResult<tokio::fs::File> {
+        ensure_file_dir_exists(path).await?;
+        OpenOptions::new()
+            .read(enable_read)
+            .write(enable_write)
+            .create_new(create_new)
+            .open(path)
+            .await
+            .map_err(|err| {
+                ObjectError::disk(format!("Failed to open file {:?}", path.to_str(),), err)
+            })
+    }
+
+    pub async fn asyncify<F, T>(f: F) -> ObjectResult<T>
+    where
+        T: Send + 'static,
+        F: FnOnce() -> ObjectResult<T> + Send + 'static,
+    {
+        spawn_blocking(f).await.map_err(|e| {
```

Comment:
It seems that `asyncify` will block one physical thread of tokio. 
How about create some independent threads to serve IO request?

Reply:
`spawn_blocking` is separated from `spawn` and is dedicated to run tokio tasks that can probably be blocked in place. It will use threads separated from threads that run normal tokio tasks(async/await tasks).

Reference: 
https://docs.rs/tokio/0.2.25/tokio/task/fn.spawn_blocking.html
https://docs.rs/tokio/0.2.25/tokio/index.html#cpu-bound-tasks-and-blocking-code